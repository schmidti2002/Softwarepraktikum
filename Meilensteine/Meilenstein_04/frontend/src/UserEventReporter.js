import { Modal, Toast } from 'bootstrap';

// Zeigt dem Nutzer Infos, Warnungen und Fehlermeldungen
// Die Klasse nutzt absichtlich kein fetch um HTML nachzuladen um m√∂glichst robust zu sein
export default class UserEventReporter {
  #toastContainer;

  #modalContainer;

  #converter;

  constructor(container) {
    try {
    // eslint-disable-next-line no-param-reassign
      container.innerHTML = '<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"></div>'
        + '<div id="modal-container"></div>';

      [this.#toastContainer] = container.getElementsByClassName('toast-container');

      this.#modalContainer = document.getElementById('modal-container');

      this.#converter = document.createElement('div');
    } catch (error) {
      // eslint-disable-next-line no-alert
      alert('Fehlerbehandlung konnte nicht initalisiert werden!');
    }
  }

  // kleiner Hack um HTML-Strings in Elemente umzuwandeln
  #strToHtmlElm(str) {
    this.#converter.innerHTML = str;
    return this.#converter.firstChild;
  }

  #spawnToast(bgColorClass, msg, autohide = false) {
    // eslint-disable-next-line operator-linebreak
    const toastHTML =
      `<div class="toast align-items-center ${bgColorClass}" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body"></div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div>`;
    const toastElm = this.#strToHtmlElm(toastHTML);
    toastElm.getElementsByClassName('toast-body')[0].textContent = msg;
    this.#toastContainer.appendChild(toastElm);
    const toast = new Toast(toastElm, { autohide });
    toastElm.addEventListener('hidden.bs.toast', () => {
      toast.dispose();
      toastElm.remove();
    });
    toast.show();
  }

  success(msg) {
    this.#spawnToast('bg-success', msg, true);
  }

  info(msg) {
    this.#spawnToast('bg-light', msg, true);
  }

  warn(msg) {
    this.#spawnToast('bg-warning', msg);
  }

  error(msg) {
    this.#spawnToast('bg-danger', msg);
  }

  fatal(msg) {
    // eslint-disable-next-line operator-linebreak
    const modalHTML =
      `<div class="modal fade" id="staticBackdrop" data-bs-backdrop="static" data-bs-keyboard="false" tabindex="-1" aria-labelledby="staticBackdropLabel" aria-hidden="true">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title" id="staticBackdropLabel">Schwerwiegender Fehler</h5>
        </div>
        <div class="modal-body"></div>
        <div class="modal-footer">
          <a href="." class="btn btn-primary">Neu laden</a>
        </div>
      </div>
    </div>
  </div>`;
    this.#modalContainer.innerHTML = modalHTML;
    const modalElm = this.#modalContainer.firstChild;
    modalElm.getElementsByClassName('modal-body')[0].textContent = msg;
    const modal = new Modal(modalElm);
    modal.show();
    window.stop();
  }
}
