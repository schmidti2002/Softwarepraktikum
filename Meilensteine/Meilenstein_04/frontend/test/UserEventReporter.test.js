import {
  beforeEach, describe, expect, test,
} from '@jest/globals';
import UserEventReporter from '../src/UserEventReporter';
import { awaitAllAsync } from './testUtils.test';

describe('UserEventReporter.js', () => {
  describe('#spawnToast using functions', () => {
    document.body.innerHTML = '<div id="mockContainer"></div>';
    let container = document.getElementById('mockContainer');
    let userEventReporter = new UserEventReporter(container);
    beforeEach(() => {
      document.body.innerHTML = '<div id="mockContainer"></div>';
      container = document.getElementById('mockContainer');
      userEventReporter = new UserEventReporter(container);
    });

    test('success() creates success-Toast', async () => {
      userEventReporter.success('TestNachricht');
      await awaitAllAsync();
      expect(container.innerHTML).toBe(
        `<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"><div class="toast align-items-center bg-success fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body"></div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div></div><div id="modal-container"></div>`,
      );
      expect(container.getElementsByClassName('toast-body')[0].innerText).toBe('TestNachricht');
    });

    test('info() creates info-Toast', async () => {
      userEventReporter.info('TestNachricht');
      await awaitAllAsync();
      expect(container.innerHTML).toBe(
        `<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"><div class="toast align-items-center bg-light fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body"></div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div></div><div id="modal-container"></div>`,
      );
    });

    test('warn() creates warn-Toast', async () => {
      userEventReporter.warn('TestNachricht');
      await awaitAllAsync();
      expect(container.innerHTML).toBe(
        `<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"><div class="toast align-items-center bg-warning fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body"></div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div></div><div id="modal-container"></div>`,
      );
      expect(container.getElementsByClassName('toast-body')[0].innerText).toBe('TestNachricht');
    });

    test('error() creates error-Toast', async () => {
      userEventReporter.error('TestNachricht');
      await awaitAllAsync();
      expect(container.innerHTML).toBe(
        `<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"><div class="toast align-items-center bg-danger fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body"></div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div></div><div id="modal-container"></div>`,
      );
      expect(container.getElementsByClassName('toast-body')[0].innerText).toBe('TestNachricht');
    });
  });

  describe('not #spawnToast using functions', () => {
    document.body.innerHTML = '<div id="mockContainer"></div>';
    let container = document.getElementById('mockContainer');
    let userEventReporter = new UserEventReporter(container);
    beforeEach(() => {
      document.body.innerHTML = '<div id="mockContainer"></div>';
      container = document.getElementById('mockContainer');
      userEventReporter = new UserEventReporter(container);
    });
    test('fatal() creates fatal-Toast', async () => {
      userEventReporter.fatal('TestNachricht');
      await awaitAllAsync();
      expect(container.innerHTML).toBe(
        `<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"></div><div id="modal-container"><div class="modal fade" id="staticBackdrop" data-bs-backdrop="static" data-bs-keyboard="false" tabindex="-1" aria-labelledby="staticBackdropLabel" aria-hidden="true" style="padding-right: 1024px;">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title" id="staticBackdropLabel">Schwerwiegender Fehler</h5>
        </div>
        <div class="modal-body"></div>
        <div class="modal-footer">
          <a href="." class="btn btn-primary">Neu laden</a>
        </div>
      </div>
    </div>
  </div></div>`,
      );
      expect(container.getElementsByClassName('modal-body')[0].innerText).toBe('TestNachricht');
    });
  });
});
