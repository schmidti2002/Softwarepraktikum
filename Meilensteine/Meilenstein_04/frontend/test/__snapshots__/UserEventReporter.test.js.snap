// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserEventReporter.js #spawnToast using functions error() creates error-Toast 1`] = `
"<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"><div class="toast align-items-center bg-danger fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body">TestNachricht</div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div></div><div id="modal-container"></div>"
`;

exports[`UserEventReporter.js #spawnToast using functions fatal() creates fatal-Toast 1`] = `
"<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"></div><div id="modal-container"><div class="modal fade" id="staticBackdrop" data-bs-backdrop="static" data-bs-keyboard="false" tabindex="-1" aria-labelledby="staticBackdropLabel" aria-hidden="true" style="padding-right: 1024px;">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title" id="staticBackdropLabel">Schwerwiegender Fehler</h5>
        </div>
        <div class="modal-body">TestNachricht</div>
        <div class="modal-footer">
          <a href="." class="btn btn-primary">Neu laden</a>
        </div>
      </div>
    </div>
  </div></div><div class="modal-backdrop fade show"></div>"
`;

exports[`UserEventReporter.js #spawnToast using functions info() creates info-Toast 1`] = `
"<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"><div class="toast align-items-center bg-light fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body">TestNachricht</div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div></div><div id="modal-container"></div>"
`;

exports[`UserEventReporter.js #spawnToast using functions multiple calls create multiple-Toast 1`] = `
"<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"><div class="toast align-items-center bg-danger fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body">TestNachricht</div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div><div class="toast align-items-center bg-light fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body">TestNachricht</div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div><div class="toast align-items-center bg-warning fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body">TestNachricht</div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div></div><div id="modal-container"></div>"
`;

exports[`UserEventReporter.js #spawnToast using functions success() creates success-Toast 1`] = `
"<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"><div class="toast align-items-center bg-success fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body">TestNachricht</div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div></div><div id="modal-container"></div>"
`;

exports[`UserEventReporter.js #spawnToast using functions warn() creates warn-Toast 1`] = `
"<div class="toast-container position-fixed bottom-0 end-0 p-3" style="z-index: 11"><div class="toast align-items-center bg-warning fade show showing" role="alert" aria-live="assertive" aria-atomic="true">
    <div class="d-flex">
      <div class="toast-body">TestNachricht</div>
      <button type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
    </div>
  </div></div><div id="modal-container"></div>"
`;
